File: src-tauri/src/audio_file.rs
use crate::error::{AppError, AppResult};
use hound::{SampleFormat, WavReader};
use std::path::Path;

pub struct AudioData {
    pub sample_rate: u32,
    pub pcm_bytes: Vec<u8>,
    pub duration_secs: f32,
}

/// Build a WAV file from raw PCM data (16-bit mono little-endian).
pub fn build_wav_bytes(sample_rate: u32, pcm_data: &[u8]) -> Vec<u8> {
    let num_channels: u16 = 1;
    let bits_per_sample: u16 = 16;
    let byte_rate = sample_rate * u32::from(num_channels) * u32::from(bits_per_sample) / 8;
    let block_align = num_channels * bits_per_sample / 8;
    let data_size = pcm_data.len() as u32;
    let file_size = 36 + data_size;

    let mut wav = Vec::with_capacity(44 + pcm_data.len());
    // RIFF header
    wav.extend_from_slice(b"RIFF");
    wav.extend_from_slice(&file_size.to_le_bytes());
    wav.extend_from_slice(b"WAVE");
    // fmt subchunk
    wav.extend_from_slice(b"fmt ");
    wav.extend_from_slice(&16u32.to_le_bytes()); // subchunk1 size
    wav.extend_from_slice(&1u16.to_le_bytes()); // audio format (PCM)
    wav.extend_from_slice(&num_channels.to_le_bytes());
    wav.extend_from_slice(&sample_rate.to_le_bytes());
    wav.extend_from_slice(&byte_rate.to_le_bytes());
    wav.extend_from_slice(&block_align.to_le_bytes());
    wav.extend_from_slice(&bits_per_sample.to_le_bytes());
    // data subchunk
    wav.extend_from_slice(b"data");
    wav.extend_from_slice(&data_size.to_le_bytes());
    wav.extend_from_slice(pcm_data);

    wav
}

pub fn read_wav_as_mono_i16(path: &Path) -> AppResult<AudioData> {
    let mut reader = WavReader::open(path).map_err(|e| AppError::Audio(e.to_string()))?;
    let spec = reader.spec();

    if spec.sample_format != SampleFormat::Int || spec.bits_per_sample != 16 {
        return Err(AppError::Audio(
            "WAVは16-bit PCMのみ対応しています".to_string(),
        ));
    }

    let channels = spec.channels as usize;
    if channels == 0 {
        return Err(AppError::Audio("channels=0".to_string()));
    }

    let mut mono: Vec<i16> = Vec::new();
    let mut frame: Vec<i16> = Vec::with_capacity(channels);

    for sample in reader.samples::<i16>() {
        let value = sample.map_err(|e| AppError::Audio(e.to_string()))?;
        frame.push(value);
        if frame.len() == channels {
            if channels == 1 {
                mono.push(frame[0]);
            } else {
                let sum: i32 = frame.iter().map(|v| i32::from(*v)).sum();
                let avg = (sum / channels as i32).clamp(i16::MIN as i32, i16::MAX as i32);
                mono.push(avg as i16);
            }
            frame.clear();
        }
    }

    let mut pcm_bytes = Vec::with_capacity(mono.len() * 2);
    for sample in mono.iter() {
        pcm_bytes.extend_from_slice(&sample.to_le_bytes());
    }

    let duration_secs = mono.len() as f32 / spec.sample_rate as f32;

    Ok(AudioData {
        sample_rate: spec.sample_rate,
        pcm_bytes,
        duration_secs,
    })
}

File: Sources/WhispCore/AudioFile.swift
import Foundation

public struct AudioData: Equatable, Sendable {
    public let sampleRate: UInt32
    public let pcmBytes: Data
    public let durationSeconds: Float

    public init(sampleRate: UInt32, pcmBytes: Data, durationSeconds: Float) {
        self.sampleRate = sampleRate
        self.pcmBytes = pcmBytes
        self.durationSeconds = durationSeconds
    }
}

public func buildWAVBytes(sampleRate: UInt32, pcmData: Data) -> Data {
    let numChannels: UInt16 = 1
    let bitsPerSample: UInt16 = 16
    let byteRate = sampleRate * UInt32(numChannels) * UInt32(bitsPerSample) / 8
    let blockAlign = numChannels * bitsPerSample / 8
    let dataSize = UInt32(pcmData.count)
    let fileSize = 36 + dataSize

    var wav = Data()
    wav.reserveCapacity(44 + pcmData.count)
    wav.append("RIFF".data(using: .ascii)!)
    wav.append(contentsOf: withUnsafeBytes(of: fileSize.littleEndian, Array.init))
    wav.append("WAVE".data(using: .ascii)!)
    wav.append("fmt ".data(using: .ascii)!)
    wav.append(contentsOf: withUnsafeBytes(of: UInt32(16).littleEndian, Array.init))
    wav.append(contentsOf: withUnsafeBytes(of: UInt16(1).littleEndian, Array.init))
    wav.append(contentsOf: withUnsafeBytes(of: numChannels.littleEndian, Array.init))
    wav.append(contentsOf: withUnsafeBytes(of: sampleRate.littleEndian, Array.init))
    wav.append(contentsOf: withUnsafeBytes(of: byteRate.littleEndian, Array.init))
    wav.append(contentsOf: withUnsafeBytes(of: blockAlign.littleEndian, Array.init))
    wav.append(contentsOf: withUnsafeBytes(of: bitsPerSample.littleEndian, Array.init))
    wav.append("data".data(using: .ascii)!)
    wav.append(contentsOf: withUnsafeBytes(of: dataSize.littleEndian, Array.init))
    wav.append(pcmData)

    return wav
}

public func parsePCM16MonoWAV(_ wavData: Data) throws -> AudioData {
    guard wavData.count >= 44 else {
        throw AppError.invalidArgument("WAVが短すぎます")
    }

    guard wavData.starts(with: Data("RIFF".utf8)),
          wavData[8..<12] == Data("WAVE".utf8)
    else {
        throw AppError.invalidArgument("WAVヘッダが不正です")
    }

    var cursor = 12
    var sampleRate: UInt32?
    var channels: UInt16?
    var bitsPerSample: UInt16?
    var pcmData: Data?

    while cursor + 8 <= wavData.count {
        let chunkID = wavData[cursor..<(cursor + 4)]
        let chunkSize = Int(readLEUInt32(wavData, offset: cursor + 4))
        let chunkDataStart = cursor + 8
        let chunkDataEnd = chunkDataStart + chunkSize

        if chunkDataEnd > wavData.count {
            break
        }

        if chunkID == Data("fmt ".utf8) {
            guard chunkSize >= 16 else {
                throw AppError.invalidArgument("fmtチャンクが不正です")
            }
            let format = readLEUInt16(wavData, offset: chunkDataStart)
            if format != 1 {
                throw AppError.invalidArgument("PCM以外のWAVは未対応です")
            }
            channels = readLEUInt16(wavData, offset: chunkDataStart + 2)
            sampleRate = readLEUInt32(wavData, offset: chunkDataStart + 4)
            bitsPerSample = readLEUInt16(wavData, offset: chunkDataStart + 14)
        } else if chunkID == Data("data".utf8) {
            pcmData = wavData.subdata(in: chunkDataStart..<chunkDataEnd)
        }

        cursor = chunkDataEnd + (chunkSize % 2)
    }

    guard let sr = sampleRate, let ch = channels, let bps = bitsPerSample, let pcm = pcmData else {
        throw AppError.invalidArgument("WAVのfmt/dataチャンクが見つかりません")
    }
    guard ch == 1 else {
        throw AppError.invalidArgument("モノラルWAVのみ対応です")
    }
    guard bps == 16 else {
        throw AppError.invalidArgument("16-bit PCM WAVのみ対応です")
    }

    let duration = pcm.count > 0 ? Float(pcm.count) / Float(sr * 2) : 0
    return AudioData(sampleRate: sr, pcmBytes: pcm, durationSeconds: duration)
}

private func readLEUInt16(_ data: Data, offset: Int) -> UInt16 {
    let value = data.withUnsafeBytes { raw in
        raw.load(fromByteOffset: offset, as: UInt16.self)
    }
    return UInt16(littleEndian: value)
}

private func readLEUInt32(_ data: Data, offset: Int) -> UInt32 {
    let value = data.withUnsafeBytes { raw in
        raw.load(fromByteOffset: offset, as: UInt32.self)
    }
    return UInt32(littleEndian: value)
}
